// sanity-plugin-mux-input/build/actions/secrets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fetchSecrets = fetchSecrets;
exports.saveSecrets = saveSecrets;
exports.testSecrets = testSecrets;

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function fetchSecrets() {
  return _client.default.fetch('*[_id == "secrets.mux"]').then(function (results) {
    var secrets = {
      token: null,
      secretKey: null
    };

    if (results.length > 0) {
      secrets.token = results[0].token || null;
      secrets.secretKey = results[0].secretKey || null;
    }

    var exists = results.length !== 0;
    return {
      secrets: secrets,
      exists: exists
    };
  });
}

function saveSecrets(token, secretKey) {
  var doc = {
    _id: 'secrets.mux',
    _type: 'mux.apiKey',
    token: token,
    secretKey: secretKey
  };
  return _client.default.createOrReplace(doc).then(function (res) {
    return {
      token: token,
      secretKey: secretKey
    };
  });
}

function testSecrets() {
  var dataset = _client.default.clientConfig.dataset;
  return _client.default.request({
    url: "/addons/mux/secrets/".concat(dataset, "/test"),
    withCredentials: true,
    method: 'GET'
  });
}
