// sanity-plugin-mux-input/build/components/Setup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = require("lodash");

var _secrets = require("../actions/secrets");

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _default4 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _default5 = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _Setup = _interopRequireDefault(require("./Setup.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var propTypes = {
  onSave: _propTypes.default.func.isRequired,
  onCancel: _propTypes.default.func.isRequired,
  secrets: _propTypes.default.shape({
    token: _propTypes.default.string,
    secretKey: _propTypes.default.string
  })
};

var MuxVideoInputSetup =
/*#__PURE__*/
function (_Component) {
  _inherits(MuxVideoInputSetup, _Component);

  _createClass(MuxVideoInputSetup, null, [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, nextState) {
      if (!nextState.secrets) {
        return null;
      }

      if (nextProps.secrets) {
        return {
          token: nextProps.secrets.token,
          secretKey: nextProps.secrets.secretKey
        };
      }

      return null;
    }
  }]);

  function MuxVideoInputSetup(props) {
    var _this;

    _classCallCheck(this, MuxVideoInputSetup);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MuxVideoInputSetup).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "tokenInputId", (0, _lodash.uniqueId)('MuxTokenInput'));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "secretKeyInputId", (0, _lodash.uniqueId)('MuxSecretInput'));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      token: null,
      secretKey: null,
      isLoading: false,
      error: null
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleTokenChanged", function (event) {
      _this.setState({
        token: event.currentTarget.value
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleSecretKeyChanged", function (event) {
      _this.setState({
        secretKey: event.currentTarget.value
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleCancel", function (event) {
      _this.props.onCancel();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleOnSubmit", function (event) {
      event.preventDefault();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleSaveToken", function () {
      var handleError = function handleError(err) {
        console.error(err); // eslint-disable-line no-console

        _this.setState({
          isLoading: false,
          error: 'Something went wrong saving the token. See console.error for more info.'
        });
      };

      _this.setState({
        isLoading: true
      });

      var token = _this.state.token || null;
      var secretKey = _this.state.secretKey || null;
      (0, _secrets.saveSecrets)(token, secretKey).then(function () {
        (0, _secrets.testSecrets)().then(function (result) {
          _this.setState({
            isLoading: false
          });

          if (result.status) {
            _this.props.onSave({
              token: token,
              secretKey: secretKey
            });

            return;
          }

          _this.setState({
            error: 'Invalid credentials'
          });
        }).catch(handleError);
      }).catch(handleError);
    });

    if (props.secrets) {
      var _props$secrets = props.secrets,
          token = _props$secrets.token,
          secretKey = _props$secrets.secretKey;
      _this.state.token = token;
      _this.state.secretKey = secretKey;
    }

    _this.firstField = _react.default.createRef();
    return _this;
  }

  _createClass(MuxVideoInputSetup, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.firstField.current.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          error = _this$state.error,
          isLoading = _this$state.isLoading;
      return _react.default.createElement("div", {
        className: _Setup.default.root
      }, _react.default.createElement("form", {
        onSubmit: this.handleOnSubmit
      }, _react.default.createElement(_default3.default, {
        legend: 'MUX API Credentials',
        level: 1,
        description: "The credentials will be stored safely in a hidden document only available to editors."
      }, _react.default.createElement(_default4.default, {
        label: "Access Token",
        labelFor: this.tokenInputId,
        level: 0,
        className: _Setup.default.formField
      }, _react.default.createElement(_default5.default, {
        id: this.tokenInputId,
        ref: this.firstField,
        onChange: this.handleTokenChanged,
        type: "text",
        value: this.state.token || ''
      })), _react.default.createElement(_default4.default, {
        label: "Secret Key",
        labelFor: this.secretKeyInputId,
        level: 0,
        className: _Setup.default.formField
      }, _react.default.createElement(_default5.default, {
        id: this.secretKeyInputId,
        onChange: this.handleSecretKeyChanged,
        type: "text",
        value: this.state.secretKey || ''
      })), _react.default.createElement("div", {
        className: _Setup.default.buttons
      }, _react.default.createElement(_default2.default, {
        loading: isLoading,
        color: "primary",
        kind: "default",
        onClick: this.handleSaveToken
      }, "Save"), _react.default.createElement(_default2.default, {
        color: "primary",
        kind: "simple",
        onClick: this.handleCancel
      }, "Cancel")), error && _react.default.createElement("p", {
        className: _Setup.default.error
      }, error)), _react.default.createElement("div", {
        className: _Setup.default.notice
      }, _react.default.createElement("p", null, "To set up a new access token, go to your", ' ', _react.default.createElement("a", {
        href: "https://dashboard.mux.com/settings/access-tokens",
        target: "_blank",
        rel: "noreferer noopener"
      }, "account on mux.com"), "."), _react.default.createElement("p", null, "The access token needs permissions: ", _react.default.createElement("strong", null, "Mux Video "), "(Full Access) and ", _react.default.createElement("strong", null, "Mux Data"), " (Read)"))));
    }
  }]);

  return MuxVideoInputSetup;
}(_react.Component);

MuxVideoInputSetup.propTypes = propTypes;
var _default = MuxVideoInputSetup;
exports.default = _default;
