// sanity-plugin-mux-input/build/components/UploadPlaceholder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _UploadPlaceholder = _interopRequireDefault(require("./UploadPlaceholder.css"));

var _uploadIcon = _interopRequireDefault(require("part:@sanity/base/upload-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var UploadPlaceholder =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(UploadPlaceholder, _React$PureComponent);

  function UploadPlaceholder() {
    _classCallCheck(this, UploadPlaceholder);

    return _possibleConstructorReturn(this, _getPrototypeOf(UploadPlaceholder).apply(this, arguments));
  }

  _createClass(UploadPlaceholder, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          invalidPaste = _this$props.invalidPaste,
          invalidFile = _this$props.invalidFile;
      return _react.default.createElement("div", null, _react.default.createElement("div", {
        className: _UploadPlaceholder.default.inner
      }, _react.default.createElement("div", {
        className: invalidFile ? [_UploadPlaceholder.default.dropFile, _UploadPlaceholder.default.invalidFile].join(' ') : _UploadPlaceholder.default.dropFile
      }, _react.default.createElement("div", {
        className: _UploadPlaceholder.default.iconContainer
      }, _react.default.createElement(_uploadIcon.default, null)), _react.default.createElement("p", {
        className: _UploadPlaceholder.default.strong
      }, _react.default.createElement("span", null, "Drop file ", invalidFile))), _react.default.createElement("div", {
        className: this.props.invalidPaste ? [_UploadPlaceholder.default.pasteFile, _UploadPlaceholder.default.invalidPaste].join(' ') : _UploadPlaceholder.default.pasteFile
      }, _react.default.createElement("div", {
        className: _UploadPlaceholder.default.iconContainer
      }, _react.default.createElement("svg", {
        xmlns: "http://www.w3.org/2000/svg",
        viewBox: "0 0 40 40",
        height: "1em",
        width: "1em"
      }, invalidPaste && _react.default.createElement("g", null, _react.default.createElement("path", {
        fill: "currentColor",
        d: "M32.38 3.76H24.8a5.41 5.41 0 0 0-10.22 0H7a3.62 3.62 0 0 0-3.62 3.62v29A3.62 3.62 0 0 0 7 40h25.38A3.62 3.62 0 0 0 36 36.38v-29a3.62 3.62 0 0 0-3.62-3.62zm-12.69 0a1.81 1.81 0 1 1-1.81 1.81 1.81 1.81 0 0 1 1.81-1.81zm12.69 32.62H7v-29h3.62v5.44h18.13V7.38h3.63z"
      }), _react.default.createElement("path", {
        fill: "currentColor",
        d: "M12.29 29.13h14.38v1.63H12.29zM12.29 18.04h4.02v1.42h-4.02zM22.84 18.04h4.02v1.42h-4.02z"
      })), !invalidPaste && _react.default.createElement("g", null, _react.default.createElement("path", {
        fill: "currentColor",
        d: "M32.49 3.62h-7.56a5.4 5.4 0 0 0-10.19 0H7.18a3.62 3.62 0 0 0-3.62 3.61v28.93a3.62 3.62 0 0 0 3.62 3.62h25.31a3.62 3.62 0 0 0 3.62-3.62V7.23a3.62 3.62 0 0 0-3.62-3.61zm-12.65 0A1.81 1.81 0 1 1 18 5.42a1.81 1.81 0 0 1 1.84-1.8zm12.65 32.54H7.18V7.23h3.61v5.43h18.09V7.23h3.61z"
      }), _react.default.createElement("path", {
        fill: "currentColor",
        d: "M20 33.15c-6.26 0-8.44-4.22-8.53-4.4l1.73-.86-.87.43.87-.44c.07.14 1.78 3.34 6.8 3.34s6.73-3.2 6.8-3.34l1.72.87c-.09.18-2.27 4.4-8.52 4.4zM12.45 18.62h4.01v3.72h-4.01zM22.98 18.62h4.01v3.72h-4.01z"
      })))), _react.default.createElement("div", null, _react.default.createElement("p", {
        className: _UploadPlaceholder.default.strong
      }, _react.default.createElement("span", null, "Paste URL"))))));
    }
  }]);

  return UploadPlaceholder;
}(_react.default.PureComponent);

exports.default = UploadPlaceholder;

_defineProperty(UploadPlaceholder, "propTypes", {
  invalidPaste: _propTypes.default.bool,
  invalidFile: _propTypes.default.bool
});
